Do not treat replica as unassigned if primary recently created and unassigned time is below a threshold.

parkertimmins opened the pull request:
Changes the way we calculate if all replicas are unassigned when primary is recently created.
Related to #107794 which did not treat replica as unassigned if primary was not yet active.

This changes extend the time when replicas are not treated as unassigned to a buffer time period after the primary has become active. This buffer time period is controlled through the setting `health.shards_availability.replica_unassigned_buffer_time`. This is only used in serverless; on stateful the behavior remains the same where  new unassigned replicas will only not be treated was unassigned while the primary is not yet active.

parkertimmins committed:
shard_availablity yellow if unassigned less than threshold

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

parkertimmins committed:
Update docs/changelog/112066.yaml

elasticsearchmachine commented:
Hi @parkertimmins, I've created a changelog YAML for you.

parkertimmins committed:
Change replica_unassigned_buffer_time default to 3s

parkertimmins committed:
Fix existing test by using buffer=0s

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/shards/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -439,10 +458,11 @@ public class ShardAllocationCounts {
         final Map<Diagnosis.Definition, Set<String>> diagnosisDefinitions = new HashMap<>();
 
         public void increment(ShardRouting routing, ClusterState state, NodesShutdownMetadata shutdowns, boolean verbose) {
+            final TimeValue replicaUnassignedThreshold = clusterService.getClusterSettings().get(REPLICA_UNASSIGNED_BUFFER_TIME);
```
This is different than the normal setup for dynamic cluster settings. Normally we have a class variable and a

```java
clusterService.getClusterSettings().addSettingsUpdateConsumer(REPLICA_UNASSIGNED_BUFFER_TIME, this::updateTheVar);
```

I think we want to do it with the update consumer, since that's the usual way for handling these settings

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/shards/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -520,18 +540,30 @@ boolean areAllShardsOfThisTypeUnavailable(ShardRouting routing, ClusterState sta
+        ShardRouting routing,
+        ClusterState state,
+        TimeValue replicaUnassignedThreshold,
+        Clock clock
```
A clock seems like overkill for this, do you think this should instead be passed `Instant.now().toEpochMilli()` and just take a `long`?

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/shards/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -520,18 +540,30 @@ boolean areAllShardsOfThisTypeUnavailable(ShardRouting routing, ClusterState sta
-        return ClusterShardHealth.getInactivePrimaryHealth(primary) == ClusterHealthStatus.YELLOW;
+
+        Optional<UnassignedInfo> ui = Optional.ofNullable(routing.unassignedInfo());
+        boolean isNewlyCreated = ui.filter(info -> info.failedAllocations() == 0)
+            .map(info -> info.lastAllocationStatus() != UnassignedInfo.AllocationStatus.DECIDERS_NO)
+            .orElse(false);
+        var cutOffTime = clock.millis() - replicaUnassignedThreshold.millis();
+        boolean hasBeenUnassignedLessThanThreshold = ui.map(info -> info.unassignedTimeMillis() > cutOffTime).orElse(false);
+        return isNewlyCreated && hasBeenUnassignedLessThanThreshold;
```
I think you can combine these:

```suggestion
        long cutOffTime = clock.millis() - replicaUnassignedThreshold.millis();
        return ui.filter(info -> info.failedAllocations() == 0)
            .map(info -> info.lastAllocationStatus() != UnassignedInfo.AllocationStatus.DECIDERS_NO)
            .map(info -> info.unassignedTimeMillis() > cutOffTime)
            .orElse(false);
```

parkertimmins committed:
Updates from review

- pass now millis instead of clock
- update setting with addSettingsUpdateConsumer
- simplify cutoff condition logic

parkertimmins committed:
remove dead code

parkertimmins committed:
more spotless

parkertimmins committed:
Fix broken test

- need to add settings to mocked ClusterService

parkertimmins commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/shards/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -108,11 +111,29 @@ public class ShardsAvailabilityHealthIndicatorService implements HealthIndicator
+     * only used in serverless, so this setting has no effect in stateless.
+     */
+    public static final Setting<TimeValue> REPLICA_UNASSIGNED_BUFFER_TIME = Setting.timeSetting(
+        "health.shard_availability.replica_unassigned_buffer_time",
```
Any opinion on the setting name?

dakrone replied:
I think the only thing to change would be to use `shards_availability` instead of `shard_availability` so that it matches the indicator name. Other than that I think it works.

parkertimmins commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/shards/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -910,11 +949,17 @@ public ShardAllocationStatus(Metadata clusterMetadata) {
 
         void addPrimary(ShardRouting routing, ClusterState state, NodesShutdownMetadata shutdowns, boolean verbose) {
-            primaries.increment(routing, state, shutdowns, verbose);
+            primaries.increment(routing, state, shutdowns, verbose, TimeValue.timeValueSeconds(0));
```
Wiring the setting through the static methods is class is a bit nasty. Decided to leave it out of `addPrimary` and just pass a dummy value.

dakrone replied:
For this, let's use `TimeValue.MINUS_ONE` as the "special" value here

dakrone reviewed (approved):
LGTM once the Serverless piece is merged and won't cause CI failures. I left one comment but nothing major, thanks Parker!

parkertimmins committed:
Small review improvments

parkertimmins committed:
Merge branch 'main' into shard-availability-replica-time-buffer
